kernel/src/util/iovec.rs

```diff
@@ -126,3 +128,189 @@ pub fn copy_iovs_from_user(start_addr: Vaddr, count: usize) -> Result<Box<[IoVec
 
     Ok(io_vecs.into_boxed_slice())
 }
+
+/// A collection of [`VmReader`]s.
+///
+/// Such readers are built from user-provided buffer, so it's always fallible.
+pub struct IoVecReader<'a>(Box<[VmReader<'a>]>);
+
+/// A collection of [`VmWriter`]s.
+///
+/// Such writerss are built from user-provided buffer, so it's always fallible.
+pub struct IoVecWriter<'a>(Box<[VmWriter<'a>]>);
+
+impl<'a> IoVecReader<'a> {
+    /// Creates a new `IoVecReaders` based on the provided `io_vecs`.
+    pub fn new(ctx: &'a Context<'a>, io_vecs: &[IoVec]) -> Result<Self> {
+        let mut readers = Vec::with_capacity(io_vecs.len());
+        let vm_space = ctx.process.root_vmar().vm_space();
+
+        for io_vec in io_vecs {
+            let reader = vm_space.reader(io_vec.base, io_vec.len)?;
+            readers.push(reader);
+        }
+
+        Ok(Self(readers.into_boxed_slice()))
+    }
+}
+
+impl<'a> IoVecWriter<'a> {
+    /// Creates a new collection of [`VmWriter`]s based on the provided `io_vecs`.
+    pub fn new(ctx: &'a Context<'a>, io_vecs: &[IoVec]) -> Result<Self> {
+        let mut writers = Vec::with_capacity(io_vecs.len());
+        let vm_space = ctx.process.root_vmar().vm_space();
+
+        for io_vec in io_vecs {
+            let writer = vm_space.writer(io_vec.base, io_vec.len)?;
+            writers.push(writer);
+        }
+
+        Ok(Self(writers.into_boxed_slice()))
+    }
+}
+
+/// Trait defining the read behavior for a collection of [`VmReader`]s.
+pub trait IoVecRead {
```

Comment:
As an alternative, have you considered/what are your feelings about implementing `IoVecRead` directly for `[VmReader]`?

Also, I'm not sure why we need to construct `[IoVec]` first and then `[VmReader]`, it doesn't seem necessary. This also introduces some unnecessary heap allocations and I don't think it makes sense.